consensus/src/dag/dag_network.rs

```diff
@@ -17,16 +28,161 @@ pub trait DAGNetworkSender: Send + Sync {
+    generator: ExponentialNumberGenerator,
+}
+
+impl Responders {
+    fn new(mut peers: Vec<Author>, initial_request_count: u32, max_request_count: u32) -> Self {
+        peers.shuffle(&mut rand::thread_rng());
+        Self {
+            peers,
+            generator: ExponentialNumberGenerator::new(initial_request_count, 2, max_request_count),
+        }
+    }
+
+    fn next_to_request(&mut self) -> Option<Vec<Author>> {
+        let count = self.generator.next().expect("should return a number");
+
+        if self.peers.is_empty() {
+            return None;
+        }
+        Some(
+            self.peers
+                .split_off(self.peers.len().checked_sub(count as usize).unwrap_or(0)),
+        )
+    }
+}
+
+pub struct RpcWithFallback {
+    responders: Responders,
+    message: DAGMessage,
+    timeout: Duration,
+
+    terminated: bool,
+    futures: Pin<
+        Box<FuturesUnordered<Pin<Box<dyn Future<Output = anyhow::Result<DAGMessage>> + Send>>>>,
+    >,
+    sender: Arc<dyn DAGNetworkSender>,
+    interval: Pin<Box<Interval>>,
+}
+
+impl RpcWithFallback {
+    pub fn new(
+        responders: Vec<Author>,
+        message: DAGMessage,
+        timeout: Duration,
+        sender: Arc<dyn DAGNetworkSender>,
+        time_service: TimeService,
+    ) -> Self {
+        Self {
+            responders: Responders::new(responders, 1, 4),
+            message,
+            timeout,
+
+            terminated: false,
+            futures: Box::pin(FuturesUnordered::new()),
+            sender,
+            interval: Box::pin(time_service.interval(timeout)),
+        }
+    }
+}
+
+async fn send_rpc(
+    sender: Arc<dyn DAGNetworkSender>,
+    peer: Author,
+    message: DAGMessage,
+    timeout: Duration,
+) -> anyhow::Result<DAGMessage> {
+    sender.send_rpc(peer, message, timeout).await
+}
+
+impl Stream for RpcWithFallback {
+    type Item = anyhow::Result<DAGMessage>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        if !self.futures.is_empty() {
+            // Check if any of the futures is ready
+            if let Poll::Ready(result) = self.futures.as_mut().poll_next(cx) {
+                return Poll::Ready(result);
+            }
+        }
+
+        // Check if the timeout has happened
+        let timeout = matches!(self.interval.as_mut().poll_next(cx), Poll::Ready(_));
+
+        if self.futures.is_empty() || timeout {
+            // try to find more responders and queue futures
+            if let Some(peers) = Pin::new(&mut self.responders).next_to_request() {
+                for peer in peers {
+                    let future = Box::pin(send_rpc(
+                        self.sender.clone(),
+                        peer,
+                        self.message.clone(),
+                        self.timeout,
+                    ));
+                    self.futures.push(future);
+                }
+            } else if self.futures.is_empty() {
```

Comment:
Related to the comment above, if we cannot rely on the timeouts in the RPC,  then futures might be pending forever and this called path will never be called.